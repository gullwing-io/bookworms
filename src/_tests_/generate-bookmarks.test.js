import { jest } from "@jest/globals";
import {
  generateImportBookmarkMarkup,
  traverseStructure,
  traverseFolders,
  traverseBookmarks,
  generateBookmarkFolderMarkup,
  generateBookmarkLinkMarkup,
  generateTimeStamp,
} from "../generate-bookmarks.js";

describe("generating bookmarks structure", () => {
  describe("generateImportBookmarkMarkup", () => {
    test("should return full structure for bookmarks", () => {
      jest
        .spyOn(Date.prototype, "toLocaleString")
        .mockReturnValue("30/09/2021, 19:43:14");
      const config = {
        label: "Bookmarks config",
        description: "Bookmarks config description",
        folders: [
          {
            label: "Folder 1",
            description: "Folder description 1",
            bookmarks: [
              {
                label: "Link 1",
                description: "Bookmark 1",
                href: "https://www.testlink1.com",
              },
            ],
          },
          {
            label: "Folder 2",
            description: "Folder description 2",
            folders: [
              {
                label: "Folder 2.1",
                description: "Folder description 2.1",
                bookmarks: [
                  {
                    label: "Link 2",
                    description: "Bookmark 2",
                    href: "https://www.testlink2.com",
                  },
                ],
              },
            ],
          },
        ],
        bookmarks: [
          {
            label: "Link 3",
            description: "Bookmark 3",
            href: "https://www.testlink3.com",
          },
        ],
      };
      expect(generateImportBookmarkMarkup(config)).toMatchSnapshot();
    });
  });
  describe("traverseStructure", () => {
    test("should return structure for browsers bookmarks", () => {
      Date.now = jest.fn(() => 1487076708000);
      const structure = {
        folders: [
          {
            label: "browsers folder 1",
            description: "browsers folder description 1",
            bookmarks: [
              {
                label: "browsers link 1",
                description: "browsers bookmark 1",
                href: "https://www.testlink1.com",
              },
            ],
          },
          {
            label: "browsers folder 2",
            description: "browsers folder description 2",
            folders: [
              {
                label: "browsers folder 2.1",
                description: "browsers folder description 2.1",
                bookmarks: [
                  {
                    label: "browsers link 2",
                    description: "browsers bookmark 2",
                    href: "https://www.testlink2.com",
                  },
                ],
              },
            ],
          },
        ],
        bookmarks: [
          {
            label: "browsers link 3",
            description: "browsers bookmark 3",
            href: "https://www.testlink3.com",
          },
        ],
      };
      expect(traverseStructure(structure, "browsers")).toMatchSnapshot();
    });
    test("should return structure for Readme bookmarks", () => {
      Date.now = jest.fn(() => 1487076708000);
      const structure = {
        folders: [
          {
            label: "Readme folder 1",
            description: "Readme folder description 1",
            bookmarks: [
              {
                label: "Readme link 1",
                description: "Readme bookmark 1",
                href: "https://www.testlink1.com",
              },
            ],
          },
          {
            label: "Readme folder 2",
            description: "Readme folder description 2",
            folders: [
              {
                label: "Readme folder 2.1",
                description: "Readme folder description 2.1",
                bookmarks: [
                  {
                    label: "Readme link 2",
                    description: "Readme bookmark 2",
                    href: "https://www.testlink2.com",
                  },
                ],
              },
            ],
          },
        ],
        bookmarks: [
          {
            label: "Readme link 3",
            description: "Readme bookmark 3",
            href: "https://www.testlink3.com",
          },
        ],
      };
      expect(traverseStructure(structure, "readme")).toMatchSnapshot();
    });
  });
  describe("traverseFolders", () => {
    test("should return simple structure for browsers folders", () => {
      Date.now = jest.fn(() => 1487076708000);
      const folders = [
        {
          label: "browsers folder 1",
          description: "browsers folder description 1",
        },
        {
          label: "browsers folder 2",
          description: "browsers folder description 2",
        },
      ];
      expect(traverseFolders(folders, "browsers")).toMatchSnapshot();
    });
    test("should return structure for nested browsers folders", () => {
      Date.now = jest.fn(() => 1487076708000);
      const folders = [
        {
          label: "browsers folder 1",
          description: "browsers folder description 1",
          folders: [
            {
              label: "browsers folder 1.1",
              description: "browsers folder description 1.1",
            },
          ],
        },
        {
          label: "browsers folder 2",
          description: "browsers folder description 2",
        },
      ];
      expect(traverseFolders(folders, "browsers")).toMatchSnapshot();
    });
    test("should return simple structure for Readme folders", () => {
      Date.now = jest.fn(() => 1487076708000);
      const folders = [
        {
          label: "Readme folder 1",
          description: "Readme folder description 1",
        },
        {
          label: "Readme folder 2",
          description: "Readme folder description 2",
        },
      ];
      expect(traverseFolders(folders, "readme")).toMatchSnapshot();
    });
    test("should return structure for nested Readme folders", () => {
      Date.now = jest.fn(() => 1487076708000);
      const folders = [
        {
          label: "Readme folder 1",
          description: "Readme folder description 1",
          folders: [
            {
              label: "Readme folder 1.1",
              description: "Readme folder description 1.1",
            },
          ],
        },
        {
          label: "Readme folder 2",
          description: "Readme folder description 2",
        },
      ];
      expect(traverseFolders(folders, "readme")).toMatchSnapshot();
    });
  });
  describe("traverseBookmarks", () => {
    test("should return structure for browsers links", () => {
      Date.now = jest.fn(() => 1487076708000);
      const bookmarks = [
        {
          label: "browsers link 1",
          description: "browsers bookmark 1",
          href: "https://www.testlink1.com",
        },
        {
          label: "browsers link 2",
          description: "browsers bookmark 2",
          href: "https://www.testlink2.com",
        },
      ];
      expect(traverseBookmarks(bookmarks, "browsers")).toMatchSnapshot();
    });
    test("should return structure for Readme links", () => {
      Date.now = jest.fn(() => 1487076708000);
      const bookmarks = [
        {
          label: "Readme link 1",
          description: "Readme bookmark 1",
          href: "https://www.testlink1.com",
        },
        {
          label: "Readme link 2",
          description: "Readme bookmark 2",
          href: "https://www.testlink2.com",
        },
      ];
      expect(traverseBookmarks(bookmarks, "readme")).toMatchSnapshot();
    });
  });

  describe("generateBookmarkFolderMarkup", () => {
    test("should return structure for browsers folder", () => {
      Date.now = jest.fn(() => 1487076708000);
      expect(
        generateBookmarkFolderMarkup(
          null,
          "browsers folder",
          "browsers folder description",
          "<div>Children HTML</div>",
          "browsers"
        )
      ).toMatchSnapshot();
    });
    test("should return structure for Readme folder", () => {
      Date.now = jest.fn(() => 1487076708000);
      expect(
        generateBookmarkFolderMarkup(
          0,
          "Readme folder",
          "Readme folder description",
          "_Children markdown_",
          "readme"
        )
      ).toMatchSnapshot();
    });
  });
  describe("generateBookmarkLinkMarkup", () => {
    test("should return structure for browsers link", () => {
      Date.now = jest.fn(() => 1487076708000);
      const bookmark = {
        label: "browsers link",
        description: "browsers bookmark",
        href: "https://www.testlink.com",
      };
      expect(
        generateBookmarkLinkMarkup(bookmark, "browsers")
      ).toMatchSnapshot();
    });
    test("should return structure for Readme link", () => {
      Date.now = jest.fn(() => 1487076708000);
      const bookmark = {
        label: "Readme link",
        description: "Readme bookmark",
        href: "https://www.testlink.com",
      };
      expect(generateBookmarkLinkMarkup(bookmark, "readme")).toMatchSnapshot();
    });
  });
  describe("generateTimeStamp", () => {
    test("should return timestamp", () => {
      Date.now = jest.fn(() => 1487076708000);
      expect(generateTimeStamp()).toEqual(1487076708000);
    });
  });
});
