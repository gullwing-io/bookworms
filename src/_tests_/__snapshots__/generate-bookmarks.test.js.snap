// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generating bookmarks structure generateBookmarkFolderMarkup should return structure for Readme folder 1`] = `
"

# Readme folder

Readme folder description
_Children markdown_"
`;

exports[`generating bookmarks structure generateBookmarkFolderMarkup should return structure for browsers folder 1`] = `
"
            <!-- browsers folder description -->
            <DT><H3 ADD_DATE="1487076708000" LAST_MODIFIED="1487076708000">browsers folder</H3>
<DL><p>
<div>Children HTML</div>
</DL><p>
    "
`;

exports[`generating bookmarks structure generateBookmarkLinkMarkup should return structure for Readme link 1`] = `
"
* [Readme link](https://www.testlink.com) - Readme bookmark
"
`;

exports[`generating bookmarks structure generateBookmarkLinkMarkup should return structure for browsers link 1`] = `
"
    <!-- browsers bookmark -->
    <DT><A HREF="https://www.testlink.com" ADD_DATE="1487076708000">browsers link</A>
    "
`;

exports[`generating bookmarks structure generateImportBookmarkMarkup should return full structure for bookmarks 1`] = `
[
  {
    "body": "
    <!-- These bookmarks were last updated on 30/09/2021, 19:43:14 using bookworms (https://github.com/thearegee/bookworms) -->
    <!DOCTYPE NETSCAPE-Bookmark-file-1>
<!-- This is an automatically generated file.
        It will be read and overwritten.
        DO NOT EDIT! -->
<META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=UTF-8">
<TITLE>Bookmarks</TITLE>
<H1>Bookmarks</H1>
<!-- Bookmarks config description -->
<DL><p>
<DT><H3 ADD_DATE="1666964654539" LAST_MODIFIED="1666964654539">Bookmarks config</H3>
<DL><p>
    
            <!-- Folder description 1 -->
            <DT><H3 ADD_DATE="1666964654539" LAST_MODIFIED="1666964654539">Folder 1</H3>
<DL><p>

    <!-- Bookmark 1 -->
    <DT><A HREF="https://www.testlink1.com" ADD_DATE="1666964654539">Link 1</A>
    
</DL><p>
    
            <!-- Folder description 2 -->
            <DT><H3 ADD_DATE="1666964654539" LAST_MODIFIED="1666964654539">Folder 2</H3>
<DL><p>

            <!-- Folder description 2.1 -->
            <DT><H3 ADD_DATE="1666964654539" LAST_MODIFIED="1666964654539">Folder 2.1</H3>
<DL><p>

    <!-- Bookmark 2 -->
    <DT><A HREF="https://www.testlink2.com" ADD_DATE="1666964654539">Link 2</A>
    
</DL><p>
    
</DL><p>
    
    <!-- Bookmark 3 -->
    <DT><A HREF="https://www.testlink3.com" ADD_DATE="1666964654539">Link 3</A>
    
</DL><p>
</DL><p>
    ",
    "filename": "browsers.html",
  },
  {
    "body": "# Bookmarks config

Bookmarks config description

* [Link 3](https://www.testlink3.com) - Bookmark 3


## Folder 1

Folder description 1

* [Link 1](https://www.testlink1.com) - Bookmark 1


## Folder 2

Folder description 2


### Folder 2.1

Folder description 2.1

* [Link 2](https://www.testlink2.com) - Bookmark 2

_These bookmarks were last updated on 30/09/2021, 19:43:14 using [Bookworms](https://github.com/thearegee/bookworms)_",
    "filename": "README.md",
  },
]
`;

exports[`generating bookmarks structure traverseBookmarks should return structure for Readme links 1`] = `
"
* [Readme link 1](https://www.testlink1.com) - Readme bookmark 1

* [Readme link 2](https://www.testlink2.com) - Readme bookmark 2
"
`;

exports[`generating bookmarks structure traverseBookmarks should return structure for browsers links 1`] = `
"
    <!-- browsers bookmark 1 -->
    <DT><A HREF="https://www.testlink1.com" ADD_DATE="1487076708000">browsers link 1</A>
    
    <!-- browsers bookmark 2 -->
    <DT><A HREF="https://www.testlink2.com" ADD_DATE="1487076708000">browsers link 2</A>
    "
`;

exports[`generating bookmarks structure traverseFolders should return simple structure for Readme folders 1`] = `
"

## Readme folder 1

Readme folder description 1


## Readme folder 2

Readme folder description 2
"
`;

exports[`generating bookmarks structure traverseFolders should return simple structure for browsers folders 1`] = `
"
            <!-- browsers folder description 1 -->
            <DT><H3 ADD_DATE="1487076708000" LAST_MODIFIED="1487076708000">browsers folder 1</H3>
<DL><p>

</DL><p>
    
            <!-- browsers folder description 2 -->
            <DT><H3 ADD_DATE="1487076708000" LAST_MODIFIED="1487076708000">browsers folder 2</H3>
<DL><p>

</DL><p>
    "
`;

exports[`generating bookmarks structure traverseFolders should return structure for nested Readme folders 1`] = `
"

## Readme folder 1

Readme folder description 1


### Readme folder 1.1

Readme folder description 1.1


## Readme folder 2

Readme folder description 2
"
`;

exports[`generating bookmarks structure traverseFolders should return structure for nested browsers folders 1`] = `
"
            <!-- browsers folder description 1 -->
            <DT><H3 ADD_DATE="1487076708000" LAST_MODIFIED="1487076708000">browsers folder 1</H3>
<DL><p>

            <!-- browsers folder description 1.1 -->
            <DT><H3 ADD_DATE="1487076708000" LAST_MODIFIED="1487076708000">browsers folder 1.1</H3>
<DL><p>

</DL><p>
    
</DL><p>
    
            <!-- browsers folder description 2 -->
            <DT><H3 ADD_DATE="1487076708000" LAST_MODIFIED="1487076708000">browsers folder 2</H3>
<DL><p>

</DL><p>
    "
`;

exports[`generating bookmarks structure traverseStructure should return structure for Readme bookmarks 1`] = `
"
* [Readme link 3](https://www.testlink3.com) - Readme bookmark 3


## Readme folder 1

Readme folder description 1

* [Readme link 1](https://www.testlink1.com) - Readme bookmark 1


## Readme folder 2

Readme folder description 2


### Readme folder 2.1

Readme folder description 2.1

* [Readme link 2](https://www.testlink2.com) - Readme bookmark 2
"
`;

exports[`generating bookmarks structure traverseStructure should return structure for browsers bookmarks 1`] = `
"
            <!-- browsers folder description 1 -->
            <DT><H3 ADD_DATE="1487076708000" LAST_MODIFIED="1487076708000">browsers folder 1</H3>
<DL><p>

    <!-- browsers bookmark 1 -->
    <DT><A HREF="https://www.testlink1.com" ADD_DATE="1487076708000">browsers link 1</A>
    
</DL><p>
    
            <!-- browsers folder description 2 -->
            <DT><H3 ADD_DATE="1487076708000" LAST_MODIFIED="1487076708000">browsers folder 2</H3>
<DL><p>

            <!-- browsers folder description 2.1 -->
            <DT><H3 ADD_DATE="1487076708000" LAST_MODIFIED="1487076708000">browsers folder 2.1</H3>
<DL><p>

    <!-- browsers bookmark 2 -->
    <DT><A HREF="https://www.testlink2.com" ADD_DATE="1487076708000">browsers link 2</A>
    
</DL><p>
    
</DL><p>
    
    <!-- browsers bookmark 3 -->
    <DT><A HREF="https://www.testlink3.com" ADD_DATE="1487076708000">browsers link 3</A>
    "
`;
